src/main.cpp

```diff
@@ -179,6 +188,165 @@ BOOL CtrlHandler(DWORD fdwCtrlType)
+// serviceWorkerThread: Keep the service alive
+
+DWORD WINAPI serviceWorkerThread(LPVOID lpParam)
+{
+	UNREFERENCED_PARAMETER(lpParam);
+
+	// Do the normal startup
+	universalMain();
+
+	return ERROR_SUCCESS;
+}
+
+
+
+
+////////////////////////////////////////////////////////////////////////////////
+// serviceSetState: Set the internal status of the service
+
+void serviceSetState(DWORD acceptedControls, DWORD newState, DWORD exitCode)
+{
+	SERVICE_STATUS serviceStatus;
+	ZeroMemory(&serviceStatus, sizeof(SERVICE_STATUS));
+	serviceStatus.dwCheckPoint = 0;
+	serviceStatus.dwControlsAccepted = acceptedControls;
+	serviceStatus.dwCurrentState = newState;
+	serviceStatus.dwServiceSpecificExitCode = 0;
+	serviceStatus.dwServiceType = SERVICE_WIN32;
+	serviceStatus.dwWaitHint = 0;
+	serviceStatus.dwWin32ExitCode = exitCode;
+
+	if (SetServiceStatus(g_StatusHandle, &serviceStatus) == FALSE)
+	{
+		LOGERROR("SetServiceStatus() failed\n");
+	}
+}
+
+
+
+
+////////////////////////////////////////////////////////////////////////////////
+// serviceCtrlHandler: Handle stop events from the Service Control Manager
+
+void WINAPI serviceCtrlHandler(DWORD CtrlCode)
+{
+	switch (CtrlCode)
+	{
+		case SERVICE_CONTROL_STOP:
+		{
+			Root.SetStopping(true);
+			serviceSetState(0, SERVICE_STOP_PENDING, 0);
+			break;
+		}
+
+		default:
+		{
+			break;
+		}
+	}
+}
+
+
+
+
+////////////////////////////////////////////////////////////////////////////////
+// serviceMain: Startup logic for running as a service
+
+void WINAPI serviceMain(DWORD argc, TCHAR *argv[])
+{
+	#if defined(_DEBUG) && defined(DEBUG_SERVICE_STARTUP)
+	Sleep(10000);
+	#endif
+	
+	char applicationFilename[MAX_PATH];
+	char applicationDirectory[MAX_PATH];
+
+	GetModuleFileName(NULL, applicationFilename, MAX_PATH);  // This binaries fill path.
+
+	// GetModuleFileName() returns the path and filename. Strip off the filename.
+	strncpy(applicationDirectory, applicationFilename, (strrchr(applicationFilename, '\\') - applicationFilename));
+	applicationDirectory[strlen(applicationDirectory)] = '\0';  // Make sure new path is null terminated
+
+	// Services are run by the SCM, and inherit its working directory - usually System32.
+	// Set the working directory to the same location as the binary.
+	SetCurrentDirectory(applicationDirectory);
+
+	g_StatusHandle = RegisterServiceCtrlHandler(SERVICE_NAME, serviceCtrlHandler);
+	
+	if (g_StatusHandle == NULL)
+	{
+		OutputDebugString("RegisterServiceCtrlHandler() failed\n");
+		serviceSetState(0, SERVICE_STOPPED, GetLastError());
+		return;
+	}
+	
+	serviceSetState(SERVICE_ACCEPT_STOP, SERVICE_RUNNING, 0);
+	
+	g_ServiceThread = CreateThread(NULL, 0, serviceWorkerThread, NULL, 0, NULL);
+	if (g_ServiceThread == NULL)
+	{
+		OutputDebugString("CreateThread() failed\n");
+		serviceSetState(0, SERVICE_STOPPED, GetLastError());
+		return;
+	}
+	WaitForSingleObject(g_ServiceThread, INFINITE);  // Wait here for a stop signal.
```

Comment:
Yes, that can be done. 

Usually with a service though, the worker thread is started to provide some monitoring of the process. Having a separate worker thread will allow cRoot to be redactored later - to add calls like beginLoop() and isQuitting() - which the service worker thread can act on instead of just waiting for the Root to return.

The second potential problem, is that service threads expect a void* parameter, used to pass data into the thread. This is a Windows specific thing, so I thought it best to leave universalMain completely cross-platform, instead of adding a parameter that would only be used on Windows.

If both of those are not of concern, I will re-write it.